src/Adapter/Order/CommandHandler/UpdateProductInOrderHandler.php

```diff
@@ -58,54 +61,65 @@ final class UpdateProductInOrderHandler extends AbstractOrderHandler implements
      */
     private $orderProductQuantityUpdater;
 
+    /**
+     * @var ContextStateManager
+     */
+    private $contextStateManager;
+
     /**
      * UpdateProductInOrderHandler constructor.
      *
      * @param OrderProductQuantityUpdater $orderProductQuantityUpdater
+     * @param ContextStateManager|null $contextStateManager
      */
     public function __construct(
-        OrderProductQuantityUpdater $orderProductQuantityUpdater
+        OrderProductQuantityUpdater $orderProductQuantityUpdater,
+        ContextStateManager $contextStateManager = null
     ) {
         $this->orderProductQuantityUpdater = $orderProductQuantityUpdater;
+        $this->contextStateManager = $contextStateManager ?: new ContextStateManager(Context::getContext());
     }
 
     /**
      * {@inheritdoc}
      */
     public function handle(UpdateProductInOrderCommand $command)
     {
-        try {
-            $order = $this->getOrder($command->getOrderId());
+        $order = $this->getOrder($command->getOrderId());
 
-            $orderDetail = new OrderDetail($command->getOrderDetailId());
-            $orderInvoice = null;
-            if (!empty($command->getOrderInvoiceId())) {
-                $orderInvoice = new OrderInvoice($command->getOrderInvoiceId());
-            }
+        $orderDetail = new OrderDetail($command->getOrderDetailId());
+        $orderInvoice = null;
+        if (!empty($command->getOrderInvoiceId())) {
+            $orderInvoice = new OrderInvoice($command->getOrderInvoiceId());
+        }
 
-            // Check fields validity
-            $this->assertProductCanBeUpdated($command, $orderDetail, $order, $orderInvoice);
-            $this->assertProductNotDuplicate($order, $orderDetail, $orderInvoice);
+        // Check fields validity
+        $this->assertProductCanBeUpdated($command, $orderDetail, $order, $orderInvoice);
+        $this->assertProductNotDuplicate($order, $orderDetail, $orderInvoice);
 
-            // Update specific price if needed
-            $product = $this->getProduct(new ProductId((int) $orderDetail->product_id), (int) $order->id_lang);
-            $combination = $this->getCombination((int) $orderDetail->product_attribute_id);
+        // Update specific price if needed
+        $product = $this->getProduct(new ProductId((int) $orderDetail->product_id), (int) $order->id_lang);
+        $combination = $this->getCombination((int) $orderDetail->product_attribute_id);
 
+        $this->contextStateManager->setShop(new Shop($order->id_shop));
```

Comment:
```suggestion
        $this->setOrderContext($contextStateManager, $order);
```

This is a recent helper added, it's safer as it overrides most of the part in the context that can go wrong (Shop, Cart, Country, Currency, ...) It is safer to override them all to avoid potential other bugs related to invalid data from context

Reply:
For more safety you should call `$this->contextStateManager->saveCurrentContext()` before setting new context overrides This is what allows the stack mechanism to work properly and avoids to mess everything up when multiple calls of `restorePreviousContext` are made.

This made me realize that `saveCurrentContext` is not called in `AbstractOrderCommandHandler::setCartContext` which is a mistake from my side!! My bad, this should be fixed to avoid any unexpected bug related to context overriding

Even though it's not strictly mandatory (it's much more important to call it in independent services like `OrderAmountUpdater`, `OrderProductQuantityUpdater`, ...) it would still be safer to use it in this method as well so that the handler classes also "protect" their overrides